### PR TITLE
fix(cloud-connect): pin rustls where the release call presents its identity (fixes #12760)

### LINKED ISSUE #12760: byoc: release() builds a rustls identity on a native-tls client, so every identity-bearing release fails to build its HTTP client
## Summary

On the `byoc` branch, `release()` builds a **rustls** client identity but never pins the rustls backend, so `reqwest` resolves the builder to **native-tls** and rejects the identity. Every call that presents an instance identity fails to construct its HTTP client:

```
ClientBuild { source: reqwest::Error { kind: Builder, source: "incompatible TLS identity type" } }
```

This is the production `release()` path, not test-only scaffolding — `release_posts_the_certificate_and_a_pop_signature_the_cloud_can_verify` fails because it calls the real function.

## Cause

`crates/runtime-cloud-connect/src/release.rs:206` constructs the identity with the rustls-only constructor:

```rust
let client_identity = reqwest::Identity::from_pem(client_pem.as_bytes())...;

let mut builder = reqwest::Client::builder()
    .timeout(Duration::from_secs(30))
    .connect_timeout(Duration::from_secs(10))
    .identity(client_identity);   // <-- no .use_rustls_tls()
```

The workspace declares `reqwest = { version = "0.13", features = ["json", "query", "rustls", ...] }` without `default-features = false`, so `default-tls` (native-tls) stays enabled alongside rustls. With both backends compiled in, a builder that never calls `.use_rustls_tls()` defaults to native-tls, and `Identity::from_pem` produced a rustls identity — hence the mismatch.

Because it depends on backend unification rather than on the code under test, it is invisible to a narrowly-scoped build and deterministic under the full release feature set the sign-off gate uses (observed `TRY 6 FAIL`, all six attempts identical).

## The fix, and the precedent for it

`release.rs` is the only place in the tree that calls `.identity()` without first pinning rustls. Every other call site already does:

- `crates/runtime/src/dataconnector/https.rs:879` — `builder = builder.use_rustls_tls().identity(identity);`
- `crates/llms/src/provider.rs:125`
- `crates/runtime/src/catalogconnector/spice_cloud.rs:129`
- `crates/runtime/tests/mtls_connector/mod.rs:205`
- `crates/runtime/tests/mtls_public/mod.rs:189`

so the change is to match them:

```rust
let mut builder = reqwest::Client::builder()
    .timeout(Duration::from_secs(30))
    .connect_timeout(Duration::from_secs(10))
    .use_rustls_tls()
    .identity(client_identity);
```

## Impact

Any `byoc`-based PR fails its sign-off on this test regardless of its own diff. #12739 changes exactly two files (`crates/runtime/src/metrics_reader.rs`, `crates/spicepod/src/lib.rs`) and neither is reachable from `runtime-cloud-connect`, but its sign-off is red on this failure and cannot go green until the base branch is fixed.

Observed in sign-off run `31191823546`.

## Note on the same run

Two cayenne failures in that run are separate and already tracked or known: `prop_sequential_memory_impl_turso` SIGABRT is #12436 (whose fix, #12561, is still stuck in the merge queue), and `prop_concurrent_mixed_dense_bloom_key_sqlite` is an intermittent.

### PR DESCRIPTION
**Pull request by grokspice**
## Summary

`release()` builds its mTLS client with `reqwest::Identity::from_pem`, which produces a **rustls** identity — but the builder never pinned a TLS backend.

`graph-rs-sdk` enables `reqwest`'s `native-tls` feature, and Cargo feature unification turns it on for the whole workspace's `reqwest 0.13.4`. With both backends compiled in, a builder that does not call `.use_rustls_tls()` resolves to native-tls and rejects the rustls identity outright:

```
ClientBuild { source: reqwest::Error { kind: Builder, source: "incompatible TLS identity type" } }
```

So every identity-bearing release failed before it reached the network. Because the trigger is backend unification rather than anything in `release.rs` itself, it is invisible to a narrowly-scoped build and only appears under the full feature set the sign-off gate uses.

`release.rs` was the only `.identity()` call site in the tree that did not pin the backend. The others already do — `crates/runtime/src/dataconnector/https.rs:879`, `crates/llms/src/provider.rs`, `crates/runtime/src/catalogconnector/spice_cloud.rs`, and the two mTLS test harnesses — so this makes it consistent rather than introducing a new convention.

Fixes #12760

## Changes

- `crates/runtime-cloud-connect/src/release.rs` — add `.use_rustls_tls()` to the release client builder, ahead of `.identity()`.
- Three regression tests that call the real `release()` rather than re-deriving its client setup.

## Verification of the mechanism

This host cannot build the crate (`build.rs` needs `protoc`, which is not installed), so the behaviour was verified in a standalone binary against the versions pinned in `Cargo.lock` — `reqwest 0.13.4`, `rcgen 0.14.8` — reproducing `release()`'s client construction exactly:

| Feature set | `.identity()` without `.use_rustls_tls()` | with `.use_rustls_tls()` |
| --- | --- | --- |
| `rustls` only | builds | builds |
| `rustls` + `native-tls` (what unification produces) | **builder error** | builds |

That is the bug and the fix, and it confirms the added tests fail on the unfixed code: under unification they get `Error::ClientBuild` where they assert `Error::Http`. The same harness ran all three assertions against a patched copy of `release()` — all pass.

Two things the probe corrected, which the tests now reflect:

- `Identity::from_pem` accepts rcgen's default ECDSA P-256 keypair, which is what `IdentityStore::generate_enrollment` produces — so the test identity is built the same way a real one is.
- `Identity::from_pem` validates PEM *structure*, not key contents: a well-formed block holding nonsense is accepted. The truncated-identity test therefore omits the key block entirely, which is what actually yields `Error::Identity`.

## Follow-up filed

#12763 — with rustls pinned here and `enroll`/`renew` still on the workspace default, the two paths validate the same endpoint against different trust roots (bundled `webpki-roots` vs the OS store). Only one configuration is affected — a self-hosted control plane that sets no `ca_cert_pem` and relies on an OS-installed CA — and it is not a regression, since `release()` previously failed in *every* configuration. Deciding the backend once for the crate is a deliberate trust-configuration change, so it is tracked separately rather than folded in here.

## Test plan

- [x] Added regression test for an identity-bearing release building its client (`release_builds_its_mtls_client_from_a_real_identity`)
- [x] Added edge case test for the extra-trust-roots branch (`release_builds_its_mtls_client_with_extra_trust_roots`)
- [x] Added edge case test for a truncated identity staying distinguishable from a client-build failure (`release_reports_a_truncated_identity_distinctly`)
- [x] `cargo fmt --all -- --check` passes
- [ ] `make lint` passes (workspace-wide fmt + full clippy) — runs as part of sign-off; `make` is not available on this host
- [ ] `make test` passes — runs as part of sign-off
- [ ] `make signoff` run after the final push (`Attestation` check is green)

The lint and test gates are the ones `make signoff` executes; they are checked off once the sign-off run reports green, not before.

### CODE REVIEW

**Comment by grokspice** [comment]
@copilot review

File: crates/runtime-cloud-connect/src/release.rs [hunk 1]
```diff
@@ -264,4 +269,98 @@ mod tests {
...
+            enc_private_key_pem: String::new(),
+            enc_public_key_pem: String::new(),
+            enc_previous_private_key_pem: String::new(),
+            cache_key_b64: String::new(),
+        }
+    }
+
+    // Port 1 is privileged and unbound, so a connection to it is refused well
+    // inside the connect timeout instead of waiting it out.
+    const UNREACHABLE_ENDPOINT: &str = "https://127.0.0.1:1";
```
[the remark this answers is not in the document]
Comment (grokspice) [reply crates/runtime-cloud-connect/src/release.rs]: Agreed — applied in ecce7cd.

The constant is now a scheme-less `not-a-url` rather than `https://127.0.0.1:1`. `release` builds the client before it hands the URL to reqwest, so the send still fails with `Error::Http` (URL parse, `kind: Builder`), which keeps the discrimination against `Error::ClientBuild` these tests actually assert on — but it now does no network I/O, so nothing listening on this host can answer in the tests' place and there is no connect timeout to wait out.

One thing I checked before making the swap: `release` runs the error through `clock_skew::looks_like_certificate_validity_failure` before classifying it. That matcher only looks for expiry/not-yet-valid substrings, so a URL parse error falls through to `Error::Http` as intended rather than being reclassified as `Error::CertificateValidity`.

File: crates/runtime-cloud-connect/src/release.rs [hunk 2]
```diff
@@ -134,9 +134,14 @@ pub async fn release(
...
         reqwest::Identity::from_pem(client_pem.as_bytes()).context(IdentitySnafu)?;
 
+    // `reqwest::Identity::from_pem` builds a rustls identity, and the workspace
+    // compiles both TLS backends in, so the backend has to be pinned to match:
+    // a builder left on the default resolves to native-tls and rejects the
+    // identity outright. Every other `.identity()` call site pins it the same way.
     let mut builder = reqwest::Client::builder()
         .timeout(Duration::from_secs(30))
         .connect_timeout(Duration::from_secs(10))
+        .use_rustls_tls()
```
[the remark this answers is not in the document]
Comment (grokspice) [reply crates/runtime-cloud-connect/src/release.rs:144]: Checked this one against the dependency graph, and the premise doesn't hold here — the doc on line 108 stays accurate, so I'm leaving it as is.

`use_rustls_tls()` does not imply bundled webpki roots. Which root store rustls gets depends on the reqwest version and its enabled features, and this workspace is on **reqwest 0.13** with `features = ["json", "query", "rustls", "gzip", "brotli", "zstd", "deflate"]`. In `Cargo.lock`, `reqwest 0.13.4` resolves `rustls-platform-verifier` and does **not** pull `webpki-roots` at all:

```
name = "reqwest"
version = "0.13.4"
dependencies = [ ..., "rustls", "rustls-pki-types", "rustls-platform-verifier", ... ]
```

(For contrast, the `reqwest 0.12.24` node in the same lockfile — pulled in transitively by other crates — *does* list `webpki-roots 1.0.7`. The 0.13 node is the one this crate compiles against.)

`rustls-platform-verifier` verifies against the **platform trust store**, which is what "production uses the system roots" describes. So pinning rustls here keeps the same trust behavior the doc already states; it changes the TLS backend, not the root store.

Worth noting the direction of the change too: before this pin, `release` didn't reach a trust decision at all — the builder resolved to native-tls and rejected the rustls identity with `incompatible TLS identity type` at `Client::builder().build()`. So this moves the call from "never worked" to "works against system roots", not from one trust store to another.

The explicit `ca_cert_pem` roots are unaffected either way — those are added via `add_root_certificate` on the same builder.

**Comment by grokspice** [comment]
Ready to merge, and it is on the critical path for the `byoc` branch — needs one maintainer approval.

**State:** all checks green, no unresolved review threads, `MERGEABLE`/`BLOCKED` solely on `REVIEW_REQUIRED`. One file, `crates/runtime-cloud-connect/src/release.rs`.

**What it unblocks.** The test this fixes — `release::tests::release_posts_the_certificate_and_a_pop_signature_the_cloud_can_verify`, failing with `ClientBuild { source: "incompatible TLS identity type" }` — is deterministic debt on the `byoc` base, and it fails the sign-off of every PR based there regardless of that PR's diff. It has now burned two multi-hour sign-off runs on #12739 alone (4h17m and 3h10m), each returning the same verdict.

The sequence out, per the diagnosis on #12739:

1. Land this on `trunk`.
2. Merge `trunk` into `byoc` (needed anyway — `byoc` is well behind).
3. #12739 then signs off against a base that has the fix.

#12769 carried the same fix stacked on #12739's own branch, which created a cycle; it has been closed in favour of this PR. Nothing further can be done from the bot side — step 1 is a maintainer approval.

**Review by lukekim — APPROVED** [review]
